src/config.rs

```diff
@@ -0,0 +1,136 @@
+use core::str::FromStr;
+
+use cbor_smol::{cbor_deserialize, cbor_serialize_bytes, cbor_serialize_extending_bytes, Bytes};
+use serde::{de::DeserializeOwned, Serialize};
+use trussed::{
+    try_syscall,
+    types::{Location, Message, Vec},
+    Client,
+};
+
+const LOCATION: Location = Location::Internal;
+const FILENAME: &str = "config";
+const FORMAT_VERSION: u8 = 0;
+
+pub trait Config: Default + PartialEq + DeserializeOwned + Serialize {
+    fn field(&mut self, key: &str) -> Option<ConfigValueMut<'_>>;
+}
+
+impl Config for () {
+    fn field(&mut self, _key: &str) -> Option<ConfigValueMut<'_>> {
+        None
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub enum ConfigValueMut<'a> {
+    Bool(&'a mut bool),
+    U8(&'a mut u8),
+}
+
+impl<'a> ConfigValueMut<'a> {
+    fn set(&mut self, value: &str) -> Result<(), ConfigError> {
+        fn set_value<T: FromStr>(target: &mut T, s: &str) -> Result<(), ConfigError> {
+            *target = s.parse().map_err(|_| ConfigError::InvalidValue)?;
+            Ok(())
+        }
+
+        match self {
+            Self::Bool(r) => set_value(*r, value),
+            Self::U8(r) => set_value(*r, value),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[repr(u8)]
+pub enum ConfigError {
+    ReadFailed = 1,
+    WriteFailed = 2,
+    NoData = 3,
+    UnsupportedFormat = 4,
+    DeserializationFailed = 5,
+    SerializationFailed = 6,
+    InvalidKey = 7,
+    InvalidValue = 8,
+    DataTooLong = 9,
+}
+
+impl From<ConfigError> for u8 {
+    fn from(error: ConfigError) -> u8 {
+        let value = error as _;
+        if value == 0 {
+            panic!("Error values may not be zero");
+        }
```

Comment:
Why is there a check here?
This can never be reached.

nit: it could also be an assert.

Reply:
Changed to an assert.  The idea is that it is easy to accidentally add a variant with discriminant zero.  The assertion is intended to catch this case early.

Reply:
Makes sense.

I would have done it through a macro + unit test. In PIV and Opcard i have a macro that implements an `all` method so that I can run tests that verify invariants on all variants of an enum in a test.
But it's also less overhead because the macros also already generates `From` implementations, so I don't need the macro just for that.

Marking as resolved.

Reply:
Looks good too.